extensions/resource-deployment/src/ui/validation/validations.ts

```diff
@@ -0,0 +1,216 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import { EOL } from 'os';
+import { throwUnless } from '../../common/utils';
+import * as loc from '../../localizedConstants';
+import { InputComponent } from '../modelViewUtils';
+import { RadioGroupLoadingComponentBuilder } from '../radioGroupLoadingComponentBuilder';
+
+export interface ValidationResult {
+	valid: boolean;
+	message?: string;
+}
+
+export type Validator = () => Promise<ValidationResult>;
+export type VariableValueGetter = (variable: string) => Promise<string | number | undefined | null>;
+export type ValueGetter = () => Promise<string | number | undefined | null>;
+
+export const enum ValidationType {
+	IsInteger = 'is_integer',
+	Regex = 'regex_match',
+	LessThanOrEqualsTo = '<=',
+	GreaterThanOrEqualsTo = '>='
+}
+
+export type ValidationInfo = RegexValidationInfo | IntegerValidationInfo | ComparisonValidationInfo;
+
+export interface ValidationInfoBase {
+	readonly type: ValidationType,
+	readonly description: string,
+}
+
+export type IntegerValidationInfo = ValidationInfoBase;
+
+export interface RegexValidationInfo extends ValidationInfoBase {
+	readonly regex: string | RegExp
+}
+
+export interface ComparisonValidationInfo extends ValidationInfoBase {
+	readonly target: string
+}
+
+export abstract class Validation {
+	private _description: string;
+
+	get description(): string {
+		return this._description;
+	}
+
+	// gets the validator for this validation object
+	abstract getValidator(): Validator;
+
+	protected getValue(): Promise<string | number | undefined | null> {
+		return this._valueGetter();
+	}
+
+	protected getVariableValue(variable: string): Promise<string | number | undefined | null> {
+		return this._variableValueGetter!(variable);
+	}
+
+	constructor(validation: ValidationInfo, protected _valueGetter: ValueGetter, protected _variableValueGetter?: VariableValueGetter) {
+		this._description = validation.description;
+	}
+}
+
+export class IntegerValidation extends Validation {
+	constructor(validation: IntegerValidationInfo, valueGetter: ValueGetter) {
+		super(validation, valueGetter);
+	}
+
+	private async isInteger(): Promise<boolean> {
+		const value = await this.getValue();
+		return (typeof value === 'string') ? Number.isInteger(parseFloat(value)) : Number.isInteger(value);
+	}
+
+	getValidator(): Validator {
+		return async () => Promise.resolve({
+			valid: await this.isInteger(),
+			message: this.description
```

Comment:
the message should be set only when the validation fails, right? this comment applies to other places as well,

Reply:
yes, but there is no harm in setting it, it actually simplifies the code. If valid the client code, should simply ignores the message whether it is set or not. If the client code makes decisions based on the absence/presence of the message value, then it is doing it wrong. Also, if the client code tries to use this value when valid is true even then it is doing it wrong.

Reply:
I don't agree, we should only set it when validation fails, code simplification should be a reason for this. it will be considered as a bug if I see an API returning error message if the validation passes.

Reply:
ok will change

Reply:
Done.